File: en/codes/ruby/chapter_sorting/counting_sort.rb
=begin
File: counting_sort.rb
Created Time: 2024-05-02
Author: Xuan Khoa Tu Nguyen (ngxktuzkai2000@gmail.com)
=end

### Counting sort ###
def counting_sort_naive(nums)
  # Simple implementation, cannot be used for sorting objects
  # 1. Count the maximum element m in the array
  m = 0
  nums.each { |num| m = [m, num].max }
  # 2. Count the occurrence of each number
  # counter[num] represents the occurrence of num
  counter = Array.new(m + 1, 0)
  nums.each { |num| counter[num] += 1 }
  # 3. Traverse counter, filling each element back into the original array nums
  i = 0
  for num in 0...(m + 1)
    (0...counter[num]).each do
      nums[i] = num
      i += 1
    end
  end
end

### Counting sort ###
def counting_sort(nums)
  # Complete implementation, can sort objects and is a stable sort
  # 1. Count the maximum element m in the array
  m = nums.max
  # 2. Count the occurrence of each number
  # counter[num] represents the occurrence of num
  counter = Array.new(m + 1, 0)
  nums.each { |num| counter[num] += 1 }
  # 3. Calculate the prefix sum of counter, converting "occurrence count" to "tail index"
  # counter[num]-1 is the last index where num appears in res
  (0...m).each { |i| counter[i + 1] += counter[i] }
  # 4. Traverse nums in reverse, fill elements into result array res
  # Initialize the array res to record results
  n = nums.length
  res = Array.new(n, 0)
  (n - 1).downto(0).each do |i|
    num = nums[i]
    res[counter[num] - 1] = num # Place num at the corresponding index
    counter[num] -= 1 # Decrement the prefix sum by 1, getting the next index to place num
  end
  # Use result array res to overwrite the original array nums
  (0...n).each { |i| nums[i] = res[i] }
end

### Driver Code ###
if __FILE__ == $0
  nums = [1, 0, 1, 2, 0, 4, 0, 2, 2, 4]

  counting_sort_naive(nums)
  puts "After counting sort (cannot sort objects), nums = #{nums}"

  nums1 = [1, 0, 1, 2, 0, 4, 0, 2, 2, 4]
  counting_sort(nums1)
  puts "After counting sort, nums1 = #{nums1}"
end

File: en/codes/rust/chapter_sorting/counting_sort.rs
/*
 * File: counting_sort.rs
 * Created Time: 2023-07-09
 * Author: night-cruise (2586447362@qq.com)
 */

use hello_algo_rust::include::print_util;

/* Counting sort */
// Simple implementation, cannot be used for sorting objects
fn counting_sort_naive(nums: &mut [i32]) {
    // 1. Count the maximum element m in the array
    let m = *nums.iter().max().unwrap();
    // 2. Count the occurrence of each number
    // counter[num] represents the occurrence of num
    let mut counter = vec![0; m as usize + 1];
    for &num in nums.iter() {
        counter[num as usize] += 1;
    }
    // 3. Traverse counter, filling each element back into the original array nums
    let mut i = 0;
    for num in 0..m + 1 {
        for _ in 0..counter[num as usize] {
            nums[i] = num;
            i += 1;
        }
    }
}

/* Counting sort */
// Complete implementation, can sort objects and is a stable sort
fn counting_sort(nums: &mut [i32]) {
    // 1. Count the maximum element m in the array
    let m = *nums.iter().max().unwrap() as usize;
    // 2. Count the occurrence of each number
    // counter[num] represents the occurrence of num
    let mut counter = vec![0; m + 1];
    for &num in nums.iter() {
        counter[num as usize] += 1;
    }
    // 3. Calculate the prefix sum of counter, converting "occurrence count" to "tail index"
    // counter[num]-1 is the last index where num appears in res
    for i in 0..m {
        counter[i + 1] += counter[i];
    }
    // 4. Traverse nums in reverse order, placing each element into the result array res
    // Initialize the array res to record results
    let n = nums.len();
    let mut res = vec![0; n];
    for i in (0..n).rev() {
        let num = nums[i];
        res[counter[num as usize] - 1] = num; // Place num at the corresponding index
        counter[num as usize] -= 1; // Decrement the prefix sum by 1, getting the next index to place num
    }
    // Use result array res to overwrite the original array nums
    nums.copy_from_slice(&res)
}

/* Driver Code */
fn main() {
    let mut nums = [1, 0, 1, 2, 0, 4, 0, 2, 2, 4];
    counting_sort_naive(&mut nums);
    print!("After counting sort (cannot sort objects) completes, nums = ");
    print_util::print_array(&nums);

    let mut nums1 = [1, 0, 1, 2, 0, 4, 0, 2, 2, 4];
    counting_sort(&mut nums1);
    print!("\nAfter counting sort, nums1 = ");
    print_util::print_array(&nums1);
}
